File: bin/checkStrings.pl
#!/usr/bin/perl
# Copyright (C) 2019-2022 The Xanado Project https://github.com/cdot/Xanado
# License MIT. See README.md at the root of this distribution for full
# copyright and license information

# Check all strings in code and HTML are reflected in en.json and translations
use strict;
use utf8;
use Path::Tiny;
use JSON;

die "Must be run from the root directory\n" unless -d "html";

binmode(STDOUT, ":utf8");
binmode(STDERR, ":utf8");

my %found;

# LOAD HTML
# Attributes scanned:
# data-i18n=
# data-i18n-placeholder=
# data-i18n-tooltip=
my $htmld;
opendir($htmld, "html");
foreach my $html (grep { /\.html$/ } readdir($htmld)) {
	my $data = path("html/$html")->slurp();
	while ($data =~ /data-i18n(?:|-placeholder|-tooltip)=(["'])(.*?)\1/g) {
		push(@{$found{$2}}, $html);
	}
}
closedir($htmld);

# LOAD JS
# Scan calls to .i18n(, grab $1
# Scan /*i18n*/ before a string, grab the string
# Scan /*i18n prefix*/ before a string, grab prefixthestring e.g.
# /*i18n namespace-$/'frood' will grab 'namespace-frood'
my $jsd;
opendir($jsd, "js");
foreach my $dir (grep { -d $_ } readdir $jsd) {
	my $jsdd;
	opendir($jsdd, $dir);
	foreach my $js (grep { /\.js$/ } readdir($jsdd)) {
		my $data = path("$dir/$js")->slurp();
		$data =~ s/[\r\n]+/ /g;
        my $report = 0;
		while ($data =~ /\.i18n\s*\(\s*(["'])(.*?)\1/g) {
			push(@{$found{$2}}, $js);
		}
		while ($data =~ /\/\*i18n(?: ([-\w]*?))?\*\/\s*(["'])(.*?)\2/g) {
			my $key = ($1 || '') . $3;
			push(@{$found{$key}}, $js);
		}
        $report = 0
	}
	closedir($jsdd);
}
closedir($jsd);

sub checkParameters {
	my ($en, $qq) = @_;
	while ($en =~ /(\$\d+)/g) {
		my $p = $1;
		my $re = "\\$p([^\\d]|\$)";
		if ($qq !~ /$re/) {
			print STDERR "\"$en\": $p not found in \"$qq\"\n";
		}
	}
}

# LOAD STRINGS
# from i18n/*.json
# If a language or qqq is passed in ARGV[0] will load that, otherwise will load
# all languages (including qqq.json, which contains documentation)
my $i18nd;
my %strings;
my @lingos = qw(en);
if (scalar(@ARGV) > 0) {
	push(@lingos, @ARGV);
} else {
	opendir($i18nd, "i18n");
	@lingos = map { $_ =~ /(.*)\.json$/; $1 } grep { /\.json$/ } readdir($i18nd)
}
foreach my $lang (@lingos) {
	my $data = decode_json(path("i18n/${lang}.json")->slurp());
	delete $data->{'@metadata'};
	$strings{$lang} = $data;
}

my $warns = 0;

foreach my $string ( sort keys %found ) {
	if (!$strings{"en"}{$string}) {
		#print "Assuming '$string' is OK\n";
		$strings{"en"}{$string} = $string;
	}
}

# CHECK STRINGS IN en.json OCCUR AT LEAST ONCE IN HTML/JS
foreach my $string (sort keys %{$strings{"en"}}) {
	if (!$found{$string}) {
		print STDERR "'$string' was found in en.json, but is not used in code\n";
		$warns++;
	}
}

# CHECK OTHER LANGUAGES
# Check that all keys in en are also in other languages.
# Check that all keys in other languages occur in en.
# qqq is treated as a language
foreach my $lang (sort keys %strings) {
	next if ($lang eq "en");
	print "Check $lang\n";
	my $titled = 0;
	foreach my $key (sort keys %{$strings{en}}) {
		if ($strings{$lang}{$key}) {
			checkParameters($key, $strings{$lang}{$key});
			next;
		}
		if (!$titled) {
			print "-------------- $lang HAS NO TRANSLATION FOR\n";
			$titled = 1;
		}
		print STDERR "\"$key\": $strings{qqq}{$key}, ";
		if ($found{$key}) {
			print STDERR "Found in ",join(", ", @{$found{$key}}),"\n";
		} else {
			print STDERR "Not found\n";
		}
		if ($strings{en}{$key} ne $key) {
			print STDERR "English is: $strings{en}{$key}\n";
		}
		$warns++;
	}

	$titled = 0;
	foreach my $key (sort keys %{$strings{$lang}}) {
		next if ($strings{"en"}{$key});
		if (!$titled) {
			print "-------------- UNUSED STRINGS IN $lang (sed script)\n";
			$titled = 1;
		}
		print STDERR "/\"$key\":/d\n";
		$warns++;
	}
}


if ($warns > 0) {
	print STDERR "$warns warnings\n";
	print STDERR "See https://github.com/wikimedia/jquery.i18n for help with string formats\n";
	print STDERR "See i18n/qqq.json for help with string meanings\n";
}

File: bin/checkStrings.js
/*Copyright (C) 2019-2022 The Xanado Project https://github.com/cdot/Xanado
License MIT. See README.md at the root of this distribution for full copyright
and license information. Author Crawford Currie http://c-dot.co.uk*/
/* eslint-env node */

/**
 * Check that strings occur in code and translations files
 */
const requirejs = require("requirejs");

requirejs.config({
  baseUrl: `${__dirname}/..`,
  nodeRequire: require
});

const readline = require('readline/promises');

requirejs([
  "fs", "path"
], (
  fs, Path
) => {

  // Maximum length for a string identifier
  const MAX_ID_LENGTH = 20;
  
  const Fs = fs.promises;

  const basePath = Path.normalize(Path.join(__dirname, ".."));

  // map string to file where it was found. Seed with @metadata which is
  // always in .json files
  const found = { "@metadata": "all .json" };
  // map file path to contents
  const fileContents = {};

  // Add string to found list
  function addString(string, file) {
    if (!found[string])
      found[string] = {};
    found[string][file] = true;
  }

  // Recursively load all files with given extension into fileContents
  // return a promise that resolves to a flat list of files loaded
  function load(file, ext, exclude) {
    if (ext.test(file) && (!exclude || !exclude.test(file))) {
      return Fs.readFile(file)
      .then(buff => fileContents[file] = buff.toString())
      .then(() => [ file ]);
    }
    return Fs.readdir(file)
    .then(files => Promise.all(
      files.map(
        f => load(Path.join(file, f), ext, exclude)))
          .then(files => files.flat()))
    .catch(e => []);
  }

  // Scan file for occurrences of re in the given files
  // and add them to found list
  function scan(files, re) {
    let m;
    for (const file of files) {
      while ((m = re.exec(fileContents[file])))
        addString(m[2], file);
    }
  }

  // check the paramers of string 'id' match in qqqString and the langString
  function checkParameters(id, qqqString, langString, mess) {
    if (/^_.*_$/.test(qqqString))
        return;
    let m, rea = /(\$\d+)/g;
    while ((m = rea.exec(qqqString))) {
      let p = m[1];
      const reb = new RegExp(`\\${p}([^\\d]|\$)`);
      if (!reb.test(langString))
        mess.push(`\t"${id}": ${p} not found in "${langString}"`);
    }
    while ((m = rea.exec(langString))) {
      let p = m[1];
      const reb = new RegExp(`\\${p}([^\\d]|\$)`);
      if (!reb.test(qqqString))
        mess.push(`\t"${id}": ${p} unexpected in "${langString}"`);
    }
  }

  // Prompt to change the id of string
  // return -2 to abort the run, -1 to ask again, 0 for no change, 1
  // if the string was changed
  async function changeLabel(lang, string, probably) {
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });
    console.log(strings[lang][string]);
    const q = `Change ID "${string}"${probably ? (' to "'+probably+'"') : ""} in ${lang}? `;
    return rl.question(q)
    .then(answer => {
      rl.close();
      switch (answer) {
      case "q": case "Q": // quit
        return -2;
      case undefined: case "": case "n": case "N":
        return 0;
      case 'y': case 'Y':
        if (probably) {
          answer = probably;
          break;
        }
      }
      if (strings[lang][answer]) {
        console.error(`${answer} is already used in ${lang}`);
        return -1; // conflict, try again
      }
      console.log(`\tChanging "${string}" to "${answer}" in ${lang}`);
      for (const lang in strings) {
        strings[lang][answer] = strings[lang][string];
        delete strings[lang][string];
      }
      const rs = string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
      const re = new RegExp(`(["'])${rs}\\1`, "g");
      const filesChanged = {};
      for (const file in fileContents) {
        const m = /i18n\/(.*)\.json$/.exec(file);
        if (m) {
          // A i18n/.json file. If the label is changed in qqq, change it everywhere.
          // If it's just changing local to a single lang, only change it there
          if (lang === "qqq" || m[1] === lang) {
            fileContents[file] = JSON.stringify(strings[m[1]], null, "  ");
            filesChanged[file] = true;
          }
        } else if (lang === "qqq" && re.test(fileContents[file])) {
          // Another file, only change the label when qqq is changing
          fileContents[file] =
          fileContents[file].replace(re, `"${answer}"`);
          filesChanged[file] = true;
        }
      }
      return Promise.all(
        Object.keys(filesChanged)
        .map(file => Fs.writeFile(file, fileContents[file])))
      .then(() => 1);
    });
  }

  async function shortenIDs() {
    for (const string in strings.qqq) {
      if (string.length > MAX_ID_LENGTH && !/^Types\./.test(strings.qqq[string])) {
        console.error(`"${string}" is too long for a label`);
        let go = -1;
        while (go === -1) {
          await changeLabel("qqq", string)
          .then(g => go = g);
        }
        if (go === -2)
          break;
      }
    }
  }

  const strings = {};

  Promise.all([
    // load with scan to extract strings
    load("html", /\.html$/)
    .then(files => scan(
      files, /data-i18n(?:|-placeholder|-tooltip)=(["'])(.*?)\1/g)),
    load("js", /\.js$/)
    .then(files => scan(files, /\.i18n\s*\(\s*(["'])(.*?)\1/g)),
    load("js", /\.js$/)
    .then(files => scan(files, /\/\*i18n\*\/\s*(["'])(.*?)\1/g)),
    // just to get fileContents
    load("test", /\.ut$/),
    load("i18n", /\.json$/, /^index\.json$/),
    Fs.readdir(Path.join(basePath, "i18n"))
    .then(lingos => Promise.all(
      lingos.filter(f => /\.json$/.test(f) && !/^index\.json$/.test(f))
      .map(lingo => Fs.readFile(Path.join(basePath, "i18n", lingo))
           .then(json => {
             const lang = lingo.replace(/\.json$/, "");
             strings[lang] = JSON.parse(json);
           })
           .catch(e => {
             console.error(`Parse error reading ${lingo}`);
             throw e;
           }))))
  ])
  .then(async () => {   
    let qqqError = false;

    // Check strings are in qqq and add to en if necessary
    for (const string of Object.keys(found).sort()) {
      if (!strings.qqq[string]) {
        console.error(`"${string}" not found in qqq`);
        strings.qqq[string] = string;
        qqqError = true;
      }
    }

    // Check strings in qqq.json occur at least once in html/js
    for (const string of Object.keys(strings.qqq)
               .filter(s => !found[s])) {
      console.error(
        `"${string}" was found in qqq, but is not used in code`);
      delete strings.qqq[string];
      qqqError = true;
    }

    if (qqqError)
      throw Error("qqq.json must be correct");

    for (const lang of Object.keys(strings).filter(l => l !== "qqq")) {
      let mess = [];

      // Check that all keys in qqq are also in other language
      for (const string of Object.keys(strings.qqq)) {
        if (!strings[lang][string])
          mess.push(`\t"${string}" : qqq "${strings.qqq[string]}" en "${strings.en[string]}"`);
      }
      if (mess.length > 0)
        console.error("----", lang, "is missing translations for:\n",
                      mess.join("\n"));

      // check that the same parameters are present in translated strings
      let messes = 0;
      for (const string of Object.keys(strings[lang])) {
        if (strings.qqq[string] && strings[lang][string]) {
          mess = [];
          checkParameters(string, strings.qqq[string], strings[lang][string], mess);
          if (mess.length > 0) {
            messes++;
            if (messes == 1)
              console.error("----", lang, "has parameter inconsistencies:");
            console.error(mess.join("\n"));
          }
        }
      }

      for (const string of Object.keys(strings[lang])) {
        if (!strings.qqq[string]) {
          console.error(`${lang}: id "${string}" was not found in qqq`);
          for (const enlabel in strings.en) {
            if (strings.en[enlabel] == string) {
              console.error(`${string} is the English translation for id ${enlabel}`);
              await changeLabel(lang, string, enlabel);
            }
          }
        }
      }

      if (lang !== "en") {
        mess = [];
        for (const id of Object.keys(strings[lang])) {
          if (strings[lang][id] == strings.en[id] && strings.en[id].length > 1)
            mess.push(`\t${id} : "${strings.en[id]}"`);
        }
        if (mess.length > 0)
          console.error(
            "----",
            lang,
            "has strings that are the same as in English\n",
            mess.join("\n"));
      }
    }
  })
  .then(() => shortenIDs());
});
